### PR TITLE
[DataLayout] Remove `clear` and `reset` methods (NFC)

### PR DESCRIPTION
**Pull request by s-barannikov**
`clear` was never necessary as it is always called on a fresh instance of the class or just before freeing an instance's memory. `reset` is effectively the same as the constructor.

### CODE REVIEW

File: llvm/lib/IR/DataLayout.cpp [hunk 1]
```diff
@@ -191,36 +191,30 @@ static const std::pair<AlignTypeEnum, LayoutAlignElem> DefaultAlignments[] = {
...
 
-  if (Error Err = parseSpecifier(Desc))
-    return report_fatal_error(std::move(Err));
+  if (Error Err = parseSpecifier(LayoutString))
+    report_fatal_error(std::move(Err));
 }
 
 DataLayout &DataLayout::operator=(const DataLayout &Other) {
-  clear();
+  // Copy everything except for LayoutMap, which will be recomputed on demand.
```
Comment (nikic) [inline llvm/lib/IR/DataLayout.cpp:217]: Shouldn't we explicitly clear it then?

Comment (s-barannikov) [reply llvm/lib/IR/DataLayout.cpp:217]: Yes, of course. 
For some reason I assumed that the operator should work the same way as a copy constructor.
Thanks for catching.

Comment (s-barannikov) [reply llvm/lib/IR/DataLayout.cpp]: Should be fixed. I had to move class declaration to make it visible to `delete`.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by maleadt** [comment]
This inadvertently caused some issues with the Julia JIT, where we used `DataLayout::reset` to replace a TargetMachine's data layout with one that marks certain address spaces as non-integral. I realize that isn't kosher, so what would be the "proper" way of creating a TargetMachine with a custom DataLayout? `createTargetMachine` doesn't have any customization.

FWIW, ignoring the mismatch between the TM's data layout and the one we use on modules causes issues down the line, e.g., when MachineFunction asserts that both match: https://github.com/llvm/llvm-project/blob/22576e2ccec60af6d27d8fd95ad3ca721b914815/llvm/lib/CodeGen/MachineFunction.cpp#L243-L245

**Comment by s-barannikov** [comment]
> we used `DataLayout::reset` to replace a TargetMachine's data layout with one that marks certain address spaces as non-integral

`TargetMachine`'s DL member is `const` and `createDataLayout()` returns a const copy of it. How did you manage to call `reset()` on a copy and get it working? If you modified `TargetMachine.h`, I guess you can also replace `reset()` with copy assignment?

**Comment by maleadt** [comment]
> `createDataLayout()` returns a const copy of it.

You're right, I misdiagnosed the issue (which did appear after upgrading to LLVM 20, so this change seemed likely).

**Pull request merged**